x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshotRepository.java

```diff
@@ -0,0 +1,189 @@
+                Iterator<ObjectObjectCursor<String, MappingMetaData>> iterator = mappings.iterator();
+                while (iterator.hasNext()) {
+                    ObjectObjectCursor<String, MappingMetaData> next = iterator.next();
+                    // we don't need to obey any routing here stuff is read-only anyway and get is disabled
+                    final String mapping = "{ \"" + next.key + "\": { \"enabled\": false, \"_meta\": " + next.value.source().string()
+                        + " } }";
+                    indexMetadataBuilder.putMapping(next.key, mapping);
+                }
+                indexMetadataBuilder.settings(Settings.builder().put(index.getSettings())
+                    .put(SOURCE_ONLY.getKey(), true)
+                    .put("index.blocks.write", true)); // read-only!
+                builder.put(indexMetadataBuilder);
+            }
+            super.initializeSnapshot(snapshotId, indices, builder.build());
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+    }
+
+    @Override
+    public void snapshotShard(IndexShard shard, Store store, SnapshotId snapshotId, IndexId indexId, IndexCommit snapshotIndexCommit,
+                              IndexShardSnapshotStatus snapshotStatus) {
+        if (shard.mapperService().documentMapper() != null // if there is no mapping this is null
+            && shard.mapperService().documentMapper().sourceMapper().isComplete() == false) {
+            throw new IllegalStateException("Can't snapshot _source only on an index that has incomplete source ie. has _source disabled " +
+                "or filters the source");
+        }
+        ShardPath shardPath = shard.shardPath();
+        Path dataPath = shardPath.getDataPath();
+        // TODO should we have a snapshot tmp directory per shard that is maintained by the system?
+        Path snapPath = dataPath.resolve(SNAPSHOT_DIR_NAME);
+        try (FSDirectory directory = new SimpleFSDirectory(snapPath)) {
+            Store tempStore = new Store(store.shardId(), store.indexSettings(), directory, new ShardLock(store.shardId()) {
+                @Override
+                protected void closeInternal() {
+                    // do nothing;
+                }
+            }, Store.OnClose.EMPTY);
+            Supplier<Query> querySupplier = shard.mapperService().hasNested() ? Queries::newNestedFilter : null;
+            // SourceOnlySnapshot will take care of soft- and hard-deletes no special casing needed here
+            SourceOnlySnapshot snapshot = new SourceOnlySnapshot(tempStore.directory(), querySupplier);
+            snapshot.syncSnapshot(snapshotIndexCommit);
+            // we will use the lucene doc ID as the seq ID so we set the local checkpoint to maxDoc with a new index UUID
+            SegmentInfos segmentInfos = store.readLastCommittedSegmentsInfo();
+            tempStore.bootstrapNewHistoryWithLocalCheckpoint(segmentInfos.totalMaxDoc());
```

Comment:
we only mark nested docs as deleted we will account for them.

Reply:
Thanks exactly my point. Say you have 3 docs with a nest doc each. In the commit, the max seq no entry says 3 for the normal index (we don't count the nested docs). In the index we'll end up restoring, the max seq# *generated* and the local checkpoint entry will be 6 (because these do account for the deleted nested docs). We need to adapt the entry in the commit just like we do for the local checkpoint.